### PR TITLE
Add transformation method to hf_causal_lm

### PR DESCRIPTION
**Pull request by irenedea**
Refactors the `ComposerHFCausalLM` class to include a `transform_model` method and make `get_peft_config` public so that subclasses can override this to transform models prior to training. 

Also creates and uses a new utility function, `set_config_overrides`.

### CODE REVIEW

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 1]
```diff
@@ -135,6 +136,17 @@ def __init__(
             should_save_peft_only=should_save_peft_only,
         )
 
+    def transform_model(self, model: PreTrainedModel) -> PreTrainedModel:
```
Comment (snarayan21) [inline llmfoundry/models/hf/hf_causal_lm.py:139]: this fn will optionally transform model, including peft adapters right

**Review by snarayan21 — APPROVED** [review]
lgtm + offline discussion

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 2]
```diff
@@ -393,12 +361,9 @@ def _autoset_attn_implementation_monkeypatch(
                 pretrained_lora_id_or_path,
             )
 
-        if prepare_for_fsdp:
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:396]: can we leave this arg here? It is part of this public api and there isn't really any reason to remove it.

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 3]
```diff
@@ -393,12 +361,9 @@ def _autoset_attn_implementation_monkeypatch(
...
-            ComposerHFCausalLM.prepare_inner_model(model, init_device)
         return model
 
-    @staticmethod
-    def _get_peft_config(peft_config_dict: Dict[str, Any]) -> 'PeftConfig':
+    def get_peft_config(self, peft_config_dict: Dict[str, Any]) -> 'PeftConfig':
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:366]: since this is public now, please add a test if one does not already exist

**Review by dakinggg — APPROVED** [review]

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 1 above)
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:139]: also please add a test for this

Comment (dakinggg) [reply llmfoundry/models/hf/hf_causal_lm.py:139]: transform is called before the peft config is created

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 2 above)
Comment (irenedea) [reply llmfoundry/models/hf/hf_causal_lm.py:396]: do you mean we should pass `prepare_for_fsdp=False` to `build_inner_model`?

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 1 above)
Comment (irenedea) [reply llmfoundry/models/hf/hf_causal_lm.py:139]: added

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 3 above)
Comment (irenedea) [reply llmfoundry/models/hf/hf_causal_lm.py:366]: added

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 2 above)
Comment (dakinggg) [reply llmfoundry/models/hf/hf_causal_lm.py:396]: yea

File: tests/models/hf/test_hf_transform.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            print('GOT PEFT CONFIG')
+            peft_config_dict['target_modules'] = ['o_proj']
+            return super().get_peft_config(peft_config_dict)
+
+    composer_model = TransformedHFCausalLM(**model_cfg)
+    model = composer_model.model
+    inner_model = maybe_get_underlying_model(model)
+
+    if peft_config:
+        print('model type', type(composer_model.model), type(composer_model))
```
Comment (dakinggg) [inline tests/models/hf/test_hf_transform.py:69]: remove

**Review by dakinggg — APPROVED** [review]

File: tests/models/hf/test_hf_transform.py (hunk 4 above)
Comment (irenedea) [inline tests/models/hf/test_hf_transform.py:69]:
```suggestion
```

File: tests/models/hf/test_hf_transform.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            with init_empty_weights():
+                model.config.num_hidden_layers = 1
+                new_model = type(model)(model.config)
+            return new_model
+
+        def get_peft_config(
+            self,
+            peft_config_dict: Dict[str, Any],
+        ) -> PeftConfig:
+            print('GOT PEFT CONFIG')
```
Comment (irenedea) [inline tests/models/hf/test_hf_transform.py:60]:
```suggestion
```

**Pull request merged**